bool b0
bool n0
bool b1
bool n1
bool b2
bool n2